src/Identifier/IdentifierNormalizer.php

```diff
@@ -0,0 +1,92 @@
+use ApiPlatform\Core\Metadata\Property\Factory\PropertyMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Property\Factory\PropertyNameCollectionFactoryInterface;
+use Symfony\Component\PropertyInfo\Type;
+
+/**
+ * Identiifer normalizer.
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class IdentifierNormalizer implements IdentifierNormalizerInterface
+{
+    private $propertyNameCollectionFactory;
+    private $propertyMetadataFactory;
+    private $compositeIdentifierNormalizer;
+    private $identifierNormalizers;
+
+    public function __construct(PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory, PropertyMetadataFactoryInterface $propertyMetadataFactory, array $identifierNormalizers)
+    {
+        $this->propertyNameCollectionFactory = $propertyNameCollectionFactory;
+        $this->propertyMetadataFactory = $propertyMetadataFactory;
+        $this->compositeIdentifierNormalizer = new CompositeIdentifierNormalizer();
+        $this->identifierNormalizers = $identifierNormalizers;
+    }
+
+    public function normalize(string $identifier, string $resourceClass): array
+    {
+        $identifiers = [];
+        $metadataIdentifiers = [];
+        $identifiersNames = [];
+        $identifiersCount = 0;
+
+        // Get metadata identifiers
+        foreach ($this->propertyNameCollectionFactory->create($resourceClass) as $propertyName) {
+            $propertyMetadata = $this->propertyMetadataFactory->create($resourceClass, $propertyName);
+
+            if (!$propertyMetadata->isIdentifier()) {
+                continue;
+            }
+
+            $type = $propertyMetadata->getType();
+            $metadataIdentifiers[$propertyName] = !$type ?: (Type::BUILTIN_TYPE_OBJECT === $type->getBuiltinType() ? $type->getClassName() : null);
+            $identifiersNames[] = $propertyName;
+            ++$identifiersCount;
+        }
+
+        // Convert to identifiers array
+        if (1 === $identifiersCount) {
+            $identifiers[$identifiersNames[0]] = $identifier;
+        } else {
+            $identifiers = $this->compositeIdentifierNormalizer->normalize($identifier);
+
+            foreach ($identifiersNames as $propertyName) {
+                if (isset($identifiers[$propertyName])) {
+                    continue;
+                }
+
+                throw new InvalidIdentifierException(sprintf('Invalid identifier "%s", "%s" has not been found.', $identifier, $propertyName));
+            }
+        }
+
+        // Normalize identifiers (DateTime, UUID etc.)
+        foreach ($identifiers as $key => $identifier) {
+            foreach ($this->identifierNormalizers as $normalizer) {
+                if ($normalizer->supportsDenormalization($identifier, $metadataIdentifiers[$key])) {
+                    try {
+                        $identifiers[$key] = $normalizer->denormalize($identifier, $metadataIdentifiers[$key]);
+                    } catch (\Exception $e) {
```

Comment:
When can this occur? Catching `\Exception` is usually broader and is a pain to debug.

Reply:
Agreed. In that case I'd have to improve the interface and we have to throw an `InvalidIdentifierException` in the denormalizer (fix #1476)